feat!: Apply permissions in frappe.qb.get_query

netchampfaris opened the pull request:
### Apply Comprehensive Permissions

- User, Role, Share Permissions
- Linked/Child Document Permissions
- Field-Level Permissions: `fields`, `filters`, `order_by`, `group_by`

### Prevent Malicious Inputs

**Fields (`SELECT`):**

Allowed
- `name`
- `` `tabUser`.`email` ``
- `field as alias`

Not allowed
- `name; DROP TABLE users`
- `field--comment`
- `invalid-field-name`
- `COUNT(*) as count` (Breaking change)

**Filters (`WHERE`):**

Allowed
- `{"name": "value"}`
- `["status", "=", "Open"]`
- `{"link_field.target_field": "X"}`
- `{"child_field.target_field": "X"}`

Not allowed
- `{"name as alias": "value"}`
- `{"tabUser.name": "value"}`
- `{"name;": "value"}`

**Group By / Order By Clauses:**

Allowed 
- `name ASC`
- `` `tabUser`.`creation` DESC ``

Not allowed
- `name as alias`
- `COUNT(name)`
- `name sideways`

### New SQL Functions syntax

#### Before
```py
frappe.qb.get_query("GP Task", fields=["sum(name) as count"])

frappe.qb.get_query("User", fields=["ifnull(first_name, 'Unknown') as safe_name"])
```

#### After
```py
frappe.qb.get_query("GP Task", fields=[{"SUM": "name", "as": "count"}])

frappe.qb.get_query(
	"User", fields=[{"IFNULL": ["first_name", "'Unknown'"], "as": "safe_name"}]
)
```
❗️This is a breaking change. This affects functions that use frappe.qb.get_query internally (`frappe.db.get_value` and `frappe.db.get_values`). I have updated usages across frappe and [erpnext](https://github.com/frappe/erpnext/pull/47761).



### Add Support for Nested Filter Conditions

The query engine now supports nested filter conditions, allowing for complex boolean logic using `AND` and `OR` operators within a structured list format.

```py

In [1]: frappe.qb.get_query("GP Task", filters=[
            ["status", "=", "Open"], "or", ["priority", "=", "High"]
        ])
Out[1]: SELECT `name` FROM `tabGP Task` WHERE `status`='Open' OR `priority`='High'

In [2]: frappe.qb.get_query(
            "GP Task",
            filters=[
                [["status", "=", "Open"], "or", ["priority", "=", "High"]],
                "and",
                ["owner", "=", "faris@frappe.io"],
            ],
        )
        
Out[2]: SELECT `name` FROM `tabGP Task` WHERE (`status`='Open' OR `priority`='High') AND `owner`='faris@frappe.io'
```

### API v2


- `/api/v2/document/<doctype>` (list endpoint) uses`frappe.qb.get_query`
`frappe.db.get_list` is replaced by `frappe.qb.get_query`
- All endpoints return `doc.as_dict()`, since `doc` only serializes fields with non-null values. `frappe-ui` needs to know when fields are set to `null` as well.

Documentation: https://docs.frappe.io/framework/user/en/api/query-builder?newWiki=1&wikiPagePatch=2tbnj6v3je&editWiki=1

---

Changes driven by this [Gameplan refactor](https://github.com/frappe/gameplan/pull/376) to use new `/api/v2`

Continuing https://github.com/frappe/frappe/pull/29082

netchampfaris committed:
fix: use frappe.qb.get_query

- static controller method get_list to modify query object

netchampfaris committed:
fix: explicitly return doc as dict

because json serializer of Document doesn't return fields with null values

netchampfaris committed:
fix: return updated doc from execute_doc_method

netchampfaris committed:
fix: print error traceback in api/v2

netchampfaris committed:
fix: rename method typing

netchampfaris committed:
fix: revert document_list return format

ankush committed:
fix: basic perm checks

netchampfaris committed:
feat: apply permissions in get_query

netchampfaris committed:
fix: print traceback instead of errprint

netchampfaris committed:
fix: add child query to allowed fields

netchampfaris committed:
fix: raise PermissionError instead of ValidationError

netchampfaris committed:
test: add permissions tests for qb.get_query

netchampfaris committed:
test: fix test_api_v2

netchampfaris committed:
test: mandatory field

netchampfaris committed:
fix: remove doc.reload

netchampfaris committed:
fix: add support for permission query conditions

netchampfaris committed:
test: simpler assertions

netchampfaris committed:
test: enable server script temporarily

netchampfaris committed:
fix: add support for AND, OR, NOT in RawCriterion

netchampfaris committed:
fix: secure query building

Add strict validation using regex for fields in SELECT, filters, GROUP BY, and ORDER BY clauses to avoid potential SQL injection risks.

Refactor field parsing and validation logic into dedicated functions.

netchampfaris committed:
refactor: Enhance field and function parsing in query engine

- Introduce `SqlFunctionParser` for robust parsing of supported SQL functions (e.g., `COUNT(*)`, `SUM(amount) as total`, `AVG(price - cost)`), replacing get_function_object and has_function.
- Refactor `DynamicTableField.parse` for improved handling of:
    - Aliases (case-insensitive `as`, quoted/unquoted).
    - `tabDocType.fieldname` notation (distinguishing child vs. main doctype refs).
    - Add validation and better error handling during parsing.
- Rewrite filter field validation (`_validate_and_prepare_filter_field`):
    - Disallow backticks (`) in filter field names.
    - Enforce specific patterns for dot notation (link/child fields only, reject `tabDoc.field`).
    - Validate character sets for simple field names.
- Update standard field parsing (`parse_string_field`, `ALLOWED_FIELD_PATTERN`, `FIELD_PARSE_REGEX`):
    - Support quoted table names potentially containing spaces (e.g., `tabTable Name`.`field`).
- Improve `parse_fields` and `_parse_single_field_item` logic:
    - Handle direct pypika `Field`/`AggregateFunction` inputs.
    - Reliably split comma-separated field strings.
```

netchampfaris committed:
test: test for invalid inputs in fields, filters, group_by, order_by

- comment out invalid tests

netchampfaris committed:
fix: restrict child table access if user has only "select" on parent

netchampfaris committed:
fix: check permlevel for fields like "link_field.fieldname"

netchampfaris committed:
fix: filtering should only be allowed on permitted fields

netchampfaris committed:
feat: add support for nested AND and OR conditions

netchampfaris committed:
chore: update new docs domain

ankush commented on frappe/database/query.py:
```diff
@@ -23,13 +26,49 @@
 
+# Pattern to validate field names in SELECT:
+# Allows: name, `name`, name as alias, `name` as alias, `table name`.`name`, `table name`.`name` as alias, table.name, table.name as alias
+ALLOWED_FIELD_PATTERN = re.compile(r"^(?:`?[\w\s-]+`?\.)?(`?\w+`?|\w+)(?:\s+as\s+\w+)?$", flags=re.ASCII)
```
I strongly feel, if we want to do this right, we should drop all kinds of regex based validations.  The maxim for security should be "_parse_, don't validate". Parse known supported values and convert them to typical code that you'd write by hand. 

Don't bypass ANY QB parameterization, this design will avoid majority of SQLi possibilities. 

e.g. `selec table.col as projection` should become `frappe.qb.from(table).select(table.col).as_(projection)`.

ankush replied:
Checked some more code. Looks like there is both parsing and validation :smile:

ankush replied:
Ig the design is mostly fine, we can get rid of risky things like functions and operators? Those should be implemented in code IMO.

ankush commented on frappe/database/query.py:
```diff
@@ -512,20 +1203,136 @@ def get_query(self, parent_names=None) -> QueryBuilder:
 		)
 
 
+class SqlFunctionParser:
```
Lets not parse functions... unnecessary risk for rarely used feature. People should use QB objects and import relevant functions in code.

sagarvora replied:
I second this. better to write a separate whitelisted method if functions are really needed.

netchampfaris committed:
feat: remove support for sql functions in fields

netchampfaris committed:
fix: harden group by and order by inputs

- only field, link_field.field, child_field.field allowed
- dont allow backticks
- add permlevel check
- add tests

netchampfaris committed:
fix: dont allow partial backticks

- add tests

netchampfaris committed:
fix: cast link fields that are int to string

- id fields should always be string
- easier to manage in typed code in frontend

netchampfaris committed:
feat: add back sql functions support with json syntax

```
fields=['user_type', {'COUNT': 'name', 'as': 'total'}]
fields=[{"IFNULL": ["first_name", "'Unknown'"], "as": "safe_name"}]
```

netchampfaris committed:
chore: update sql function usage

netchampfaris committed:
chore: update sql function usage

netchampfaris committed:
fix: revert get_all change

get_all doesn't use get_query

netchampfaris committed:
fix: typing and error handling

- add typing and comprehensive docstring
- error handling for custom get_list methods in doctype controllers

netchampfaris committed:
chore: remove unused code

netchampfaris commented:
Docs: https://docs.frappe.io/framework/get_query
